src/dict/dsl_details.cc

```diff
@@ -831,35 +832,35 @@ DslScanner::DslScanner( string const & fileName ):
 
   bool needExactEncoding = false;
 
-  QByteArray ba = QByteArray::fromRawData( (const char *)firstBytes, 50 );
-  codec         = QTextCodec::codecForUtfText( ba, nullptr );
-  if ( !codec ) {
-    // the encoding has no bom.
-    // check the first char # (0x23).
-    auto hashTag = 0x0023;
-
-    auto uci = qFromUnaligned< uint32_t >( firstBytes );
-    if ( uci == qToBigEndian( hashTag ) ) {
-      codec = QTextCodec::codecForMib( 1018 ); // utf-32 be
-    }
-    else if ( uci == qToLittleEndian( hashTag ) ) {
-      codec = QTextCodec::codecForMib( 1019 ); // utf-32 le
-    }
-    else {
-      auto uc = qFromUnaligned< uint16_t >( firstBytes );
-      if ( uc == qToBigEndian( uint16_t( hashTag ) ) ) {
-        codec = QTextCodec::codecForMib( 1013 ); // utf16 be
-      }
-      else if ( uc == qToLittleEndian( uint16_t( hashTag ) ) ) {
-        codec = QTextCodec::codecForMib( 1014 ); // utf16 le
-      }
-      else {
-        //default encoding
-        codec = QTextCodec::codecForName( "UTF-8" );
-      }
+  // Note that .dsl format always starts with "#NAME"
+  if ( auto guessedEncoding = QStringConverter::encodingForData( { firstBytes, N }, '#' );
+       guessedEncoding.has_value() ) {
+    switch ( guessedEncoding.value() ) {
+      case QStringConverter::Utf8:
+        encoding = Utf8::Utf8;
+        break;
+      case QStringConverter::Utf16LE:
+        encoding = Utf8::Utf16LE;
+        break;
+      case QStringConverter::Utf16BE:
+        encoding = Utf8::Utf16BE;
+        break;
+      case QStringConverter::Utf32LE:
+        encoding = Utf8::Utf16LE;
+        break;
+      case QStringConverter::Utf32BE:
+        encoding = Utf8::Utf32BE;
+        break;
+      default:
+        break;
```

Comment:
Not sure about here, the default of `encoding` in the original GD encoding is actually CP1251?CP1252?

In original GD, `encoding` is initialized as 1252, then proceed to set 1251 🤮   

https://github.com/goldendict/goldendict/blob/97d879edc89ab6fd7e7099a466cc10924d7c704b/dsl_details.cc#L872C13-L872C24

https://github.com/goldendict/goldendict/blob/97d879edc89ab6fd7e7099a466cc10924d7c704b/dsl_details.cc#L929

We changed the default to utf-8 since the addition of better encoding guessing, but there is no complain at all.

Reply:
I just changed the init value of encoding to utf-8 to stay sane 😅

Reply:
UTF-8/Windows-1251/Windows-1252 all includes the basic ASCII, so this doesn't really matters. Though, I am not sure how will QString handles invalid characters.